### PR TITLE
CFE-177 - Admission Control for Infrastructure CR

### PR DESCRIPTION
**Pull request by sherine-k**
**What type of PR is this?**

/kind feature
**What this PR does / why we need it:**

This PR implements the API admission control for the [Custom AWS Tags](https://github.com/openshift/enhancements/blob/master/enhancements/api-review/custom-tags-aws.md)
It validates that tag keys added in Infrastructure CR's spec.platformSpec.resourceTags :
*  do not contain `kubernetes.io`
* do not contain `openshift.io`
* do not start with `aws:`
 
**Which issue(s) this PR fixes:**

Fixes [CFE-177](https://issues.redhat.com/browse/CFE-177)
**Special notes for your reviewer:**
**Does this PR introduce a user-facing change?**

This PR will help show the user the necessary errors if any when trying to add or update aws custom tags to the infrastructure CR.

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:**

https://github.com/openshift/enhancements/blob/master/enhancements/api-review/custom-tags-aws.md
Depends on : https://github.com/openshift/api/pull/1064 => remove replace in go.mod once merged

### CODE REVIEW

**Comment by sherine-k** [comment]
/assign @sanchezl 
cc @josefkarasek @TrilokGeer

File: openshift-kube-apiserver/admission/customresourcevalidation/infrastructure/validate_infrastructure.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+)
+
+const (
+	PluginName = "config.openshift.io/ValidateInfrastructure"
+)
+
+type infrastructureV1 struct{}
+
+// kubernetesNamespaceRegex is used to check that a tag key is not in the kubernetes.io namespace.
+var kubernetesNamespaceRegex = regexp.MustCompile(`^([^/]*\.)?kubernetes.io/`)
```
Comment (deads2k) [inline openshift-kube-apiserver/admission/customresourcevalidation/infrastructure/validate_infrastructure.go:24]: does this also stop "kubernetes.io", "kubernetes.io/"?  Can I get a unit test?

**Comment by deads2k** [comment]
adding an admission plugin directly to openshift/kubernetes for infrastructure is allowed because it has to be present before we have the ability to have an admission webhook

File: openshift-kube-apiserver/admission/customresourcevalidation/infrastructure/validate_infrastructure.go (hunk 1 above)
Comment (sherine-k) [reply openshift-kube-apiserver/admission/customresourcevalidation/infrastructure/validate_infrastructure.go:24]: Hi @deads2k 
This just stops kubernetes.io/, from they way I understood our needs.
I can for sure update it to catch kubernetes.io and openshift.io as well

Comment (sanchezl) [reply openshift-kube-apiserver/admission/customresourcevalidation/infrastructure/validate_infrastructure.go]: It is unclear to me what the validation is supposed to be: 
* The CRD and the enhancement only state that the tag name cannot contain a `kubernetes.io` or `openshift.io` **prefix**.
* The commit text states that the tag name cannot contain a `kubernetes.io` or `openshift.io` **suffix**.
* The PR description state that the tag name cannot contain a `kubernetes.io` or `openshift.io` **at all**.

Comment (sherine-k) [reply openshift-kube-apiserver/admission/customresourcevalidation/infrastructure/validate_infrastructure.go]: Hi @sanchezl 
Thanks for reviewing!
You are right about this, it reflects the discussions we've had with the api team the last few days.
The final decision was to forbid tag names containing  `kubernetes.io` or `openshift.io`.
I can update the commit text when squashing commits
@TrilokGeer do we need to update the enhancement?

**Review by sanchezl — CHANGES_REQUESTED** [review]
Please split the vendor/staging update into a separate commit.

File: openshift-kube-apiserver/admission/customresourcevalidation/infrastructure/validate_infrastructure.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+type infrastructureV1 struct{}
+
+var reservedRegexes = []*regexp.Regexp{
+	//should not contain 'kubernetes.io'
+	regexp.MustCompile(`^([^/]*\.)?kubernetes.io`),
+	//should not contain 'openshift.io'
+	regexp.MustCompile(`^([^/]*\.)?openshift.io`),
+	//should not start with 'aws:'
+	regexp.MustCompile(`^aws:.*`),
```
[the remark this answers is not in the document]
Comment (sherine-k) [reply openshift-kube-apiserver/admission/customresourcevalidation/infrastructure/validate_infrastructure.go]: Suggestion committed

File: openshift-kube-apiserver/admission/customresourcevalidation/infrastructure/validate_infrastructure.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+	errs := field.ErrorList{}
+	if len(infra.Spec.PlatformSpec.AWS.ResourceTags) > 0 {
+		var unique map[string]string = make(map[string]string)
+		for index, tag := range infra.Spec.PlatformSpec.AWS.ResourceTags {
+			if _, ok := unique[tag.Key]; ok {
+				// second occurrence - is not unique
+				err := field.Duplicate(
+					field.NewPath("spec", getResourceTagsPath(index)...),
+					tag.Key)
+				errs = append(errs, err)
```
[the remark this answers is not in the document]
Comment (sherine-k) [reply openshift-kube-apiserver/admission/customresourcevalidation/infrastructure/validate_infrastructure.go:137]: Suggestion committed. 
No need to re-add the key after that. Thanks @sanchezl

File: openshift-kube-apiserver/admission/customresourcevalidation/infrastructure/validate_infrastructure.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+func getResourceTagsPath(index int) []string {
+	pathSlice := []string{"platformSpec", "AWS", "resourceTags"}
+	item := fmt.Sprintf("[%d]", index)
+	pathSlice = append(pathSlice, item)
+	return pathSlice
+}
+
+func validateTagKeyPrefix(reservedPrefix *regexp.Regexp, index int, key string) *field.Error {
```
[the remark this answers is not in the document]
Comment (sherine-k) [reply openshift-kube-apiserver/admission/customresourcevalidation/infrastructure/validate_infrastructure.go]: Suggestion committed

File: openshift-kube-apiserver/admission/customresourcevalidation/infrastructure/validate_infrastructure.go (hunk 1 above)
Comment (sherine-k) [reply openshift-kube-apiserver/admission/customresourcevalidation/infrastructure/validate_infrastructure.go]: Git commit message updated

**Comment by sherine-k** [comment]
day2 operations are not in scope for now
Closing

**Pull request closed without merging**